consensus/hotstuff/voteaggregator/vote_aggregator_v2_test.go

```diff
@@ -0,0 +1,251 @@
+	clr.On("AddVote", secondVote).Return(model.NewDoubleVoteErrorf(firstVote, secondVote, ""))
+	s.notifier.On("OnDoubleVotingDetected", firstVote, secondVote).Return(nil)
+
+	err := s.aggregator.processQueuedVote(secondVote)
+	// double voting should be handled internally and reported to notifier
+	require.NoError(s.T(), err)
+	s.notifier.AssertExpectations(s.T())
+}
+
+// TestVoteProcessing_DecreasingPruningHeightError tests that adding vote while pruning in parallel thread doesn't
+// trigger an error
+func (s *VoteAggregatorV2TestSuite) TestVoteProcessing_DecreasingPruningHeightError() {
+	staleView := uint64(1000)
+	*s.collectors = mocks.VoteCollectors{}
+	s.collectors.On("GetOrCreateCollector", staleView).Return(nil, false, mempool.NewDecreasingPruningHeightError(""))
+	err := s.aggregator.processQueuedVote(unittest.VoteFixture(unittest.WithVoteView(staleView)))
+	require.NoError(s.T(), err)
+	s.collectors.AssertExpectations(s.T())
+}
+
+// TestVoteProcessing_ProcessingExceptions tests that all unexpected errors are propagated to caller
+func (s *VoteAggregatorV2TestSuite) TestVoteProcessing_ProcessingExceptions() {
+	vote := unittest.VoteFixture()
+	// failing to create collector is a critical error
+	err := s.aggregator.processQueuedVote(vote)
+	require.ErrorIs(s.T(), err, voteCollectorsFactoryError)
+
+	view := uint64(1000)
+	vote = unittest.VoteFixture(unittest.WithVoteView(view))
+	clr := s.prepareMockedCollector(view)
+	expectedError := errors.New("processing-error")
+	clr.On("AddVote", vote).Return(expectedError)
+	err = s.aggregator.processQueuedVote(vote)
+	require.ErrorIs(s.T(), err, expectedError)
+}
+
+// TestInvalidBlock tests that InvalidBlock reports all previously processed votes to hotstuff.Consumer
+// By definition only votes for invalid proposal has to be reported
+func (s *VoteAggregatorV2TestSuite) TestInvalidBlock() {
+	view := uint64(1000)
+	proposal := helper.MakeProposal(s.T(), helper.WithBlock(helper.MakeBlock(s.T(), helper.WithBlockView(view))))
+
+	votes := 10
+	consumedVotes := make([]*model.Vote, 0, votes*2)
+	// generate votes for two proposals
+	otherBlockID := unittest.IdentifierFixture()
+	for i := 0; i < votes; i++ {
+		vote := unittest.VoteFixture(unittest.WithVoteView(view),
+			unittest.WithVoteBlockID(proposal.Block.BlockID))
+		otherVote := unittest.VoteFixture(unittest.WithVoteView(view),
+			unittest.WithVoteBlockID(otherBlockID))
+		consumedVotes = append(consumedVotes, vote, otherVote)
+
+		// only votes for one of the proposals should be reported
+		s.notifier.On("OnVoteForInvalidBlockDetected", vote, proposal)
+	}
+
+	clr := s.prepareMockedCollector(view)
+	clr.On("RegisterVoteConsumer", mock.Anything).Return(nil).Run(func(args mock.Arguments) {
+		callback := args.Get(0).(hotstuff.VoteConsumer)
+		for _, vote := range consumedVotes {
+			callback(vote)
+		}
+	})
+
+	err := s.aggregator.InvalidBlock(proposal)
+	require.NoError(s.T(), err)
+	s.notifier.AssertExpectations(s.T())
```

Comment:
can we verify that future votes after the call to `InvalidBlock` will also be reported as invalid vote?